numpy/core/einsumfunc.py

```diff
@@ -0,0 +1,718 @@
+        memory_arg = min(memory_arg, out_size)
+        path = _path_greedy(input_sets, output_set, dimension_dict, memory_arg)
+    elif path_arg == "optimal":
+        path = _path_optimal(input_sets, output_set, dimension_dict, memory_arg)
+    else:
+        raise KeyError("Path name %s not found", path_arg)
+
+    cost_list, scale_list, size_list = [], [], []
+    contraction_list = []
+
+    # Build contraction tuple (positions, gemm, einsum_str, remaining)
+    for cnum, contract_inds in enumerate(path):
+        # Make sure we remove inds from right to left
+        contract_inds = tuple(sorted(list(contract_inds), reverse=True))
+
+        contract = _find_contraction(contract_inds, input_sets, output_set)
+        out_inds, input_sets, idx_removed, idx_contract = contract
+
+        cost = _compute_size_by_dict(idx_contract, dimension_dict)
+        if idx_removed:
+            cost *= 2
+        cost_list.append(cost)
+        scale_list.append(len(idx_contract))
+        size_list.append(_compute_size_by_dict(out_inds, dimension_dict))
+
+        tmp_inputs = []
+        for x in contract_inds:
+            tmp_inputs.append(input_list.pop(x))
+
+        # Last contraction
+        if (cnum - len(path)) == -1:
+            idx_result = output_subscript
+        else:
+            sort_result = [(dimension_dict[ind], ind) for ind in out_inds]
+            idx_result = "".join([x[1] for x in sorted(sort_result)])
+
+        input_list.append(idx_result)
+        einsum_str = ",".join(tmp_inputs) + "->" + idx_result
+        contraction = (contract_inds, False, einsum_str, input_list[:])
+        contraction_list.append(contraction)
+
+    opt_cost = sum(cost_list)
+
+    if einsum_call_arg:
+        return (operands, contraction_list)
+
+    # Return the path along with a nice string representation
+    overall_contraction = input_subscripts + "->" + output_subscript
+    header = ("scaling", "BLAS", "current", "remaining")
+
+    path_print  = "  Complete contraction:  %s\n" % overall_contraction
+    path_print += "         Naive scaling:  %2d\n" % len(indices)
+    path_print += "      Naive FLOP count:  %.3e\n" % naive_cost
+    path_print += "  Optimized FLOP count:  %.3e\n" % opt_cost
+    path_print += "   Theoretical speedup:  %3.3f\n" % (naive_cost / float(opt_cost))
+    path_print += "  Largest intermediate:  %.3e elements\n" % max(size_list) 
+    path_print += "-" * 80 + "\n"
+    path_print += "%6s %6s %24s %40s\n" % header
+    path_print += "-" * 80
+
+    for n, contraction in enumerate(contraction_list):
+        inds, gemm, einsum_str, remaining = contraction
+        remaining_str = ",".join(remaining) + "->" + output_subscript
+        path_run = (scale_list[n], gemm, einsum_str, remaining_str)
+        path_print += "\n%4d    %6s %24s %40s" % path_run
+
+    return (path, path_print)
+
+# Rewrite einsum to handle different cases
+def einsum(*operands, **kwargs):
```

Comment:
Hmm, looks like that `einsum` signature was due to the documentation rather than the C function and is something implemented at the c level, so there isn't much choice here. I guess the question becomes whether to check for unexpected keyword arguments.